src/io_utils.c

```diff
@@ -87,24 +91,73 @@ sscg_normalize_path (TALLOC_CTX *mem_ctx,
                      char **_normalized_path)
 {
   int ret;
+  char *path_copy = NULL;
+  char *dir_copy = NULL;
+  char *dir_part = NULL;
+  char *file_part = NULL;
+  char *resolved_dir = NULL;
   char *normalized_path = NULL;
 
   TALLOC_CTX *tmp_ctx = talloc_new (NULL);
   CHECK_MEM (tmp_ctx);
 
-  normalized_path = talloc_zero_array (tmp_ctx, char, PATH_MAX);
-  CHECK_MEM (normalized_path);
+  /* Validate input */
+  if (path == NULL || path[0] == '\0')
+    {
+      ret = EINVAL;
+      goto done;
+    }
 
-  ret = make_normalized_absolute_path (normalized_path, PATH_MAX, path);
-  CHECK_OK (ret);
+  /* Make copies for dirname() and basename() since they modify the input */
+  path_copy = talloc_strdup (tmp_ctx, path);
+  CHECK_MEM (path_copy);
 
-  *_normalized_path = talloc_strdup (mem_ctx, normalized_path);
-  CHECK_MEM (*_normalized_path);
+  dir_copy = talloc_strdup (tmp_ctx, path);
+  CHECK_MEM (dir_copy);
+
+  /* Extract directory and filename components */
+  dir_part = dirname (dir_copy);
+  file_part = basename (path_copy);
+
+  /* Allocate buffer for realpath() */
+  resolved_dir = talloc_zero_array (tmp_ctx, char, PATH_MAX);
+  CHECK_MEM (resolved_dir);
+
+  /* Normalize the directory path */
+  if (realpath (dir_part, resolved_dir) == NULL)
```

Comment:
@sgallagher, why do you need to avoid normalization of a file part?
It's explicit in "Test 4", but why?